prusti-viper/src/encoder/backtranslation.rs

```diff
@@ -0,0 +1,457 @@
+        let return_local = Local::from(mir::Local::from_usize(0));
+        //make sure
+        let result_to_process = if local_variable_manager.is_return(return_local){
+            //open question: what would be the right span for return type? None?
+            let rust_name = String::from("result");
+            let vir_name = if !backtranslator.is_pure {
+                local_variable_manager.get_name(return_local)
+            } else {
+                String::from("Result()")
+            };
+            let typ = local_variable_manager.get_type(return_local);
+            Some((rust_name, vir_name, typ))
+        } else {
+            //this case should probably never occur
+            None
+        };
+        
+
+        //now map those needed:
+        let mut entries = HashMap::new();
+        let mut args = HashMap::new();
+
+
+        let result = if !backtranslator.is_pure {
+            if backtranslator.silicon_counterexample.old_models.contains_key(&old_label){ 
+                let silicon_arg_entries = &backtranslator.silicon_counterexample.old_models
+                    .get(&old_label)
+                    .unwrap()
+                    .entries;
+                
+        
+                for (rust_name, span, vir_name, typ) in args_to_process {
+                    let opt_entry = silicon_arg_entries.get(&vir_name);
+                    let entry = backtranslator.backtranslate_entry(
+                        typ, 
+                        opt_entry,
+                        vir_name,
+                        silicon_arg_entries,
+                    );
+                    args.insert((rust_name, span), entry);
+                }
+            }
+
+            let silicon_final_entries = if last_label.is_some() {
+                &backtranslator.silicon_counterexample.old_models
+                    .get(last_label.unwrap())
+                    .unwrap()
+                    .entries
+                } else {
+                    &backtranslator.silicon_counterexample.model.entries 
+                };
+
+            for (rust_name, span, vir_name, typ) in entries_to_process {
+                let opt_entry = silicon_final_entries.get(&vir_name);
+                let entry = backtranslator.backtranslate_entry(
+                    typ,
+                    opt_entry, 
+                    vir_name,
+                    silicon_final_entries,
+                );
+                entries.insert((rust_name, span), entry);
+            }
+            match result_to_process {
+                None => Entry::Unit,
+                Some((rust_name, vir_name, typ)) => {
+                    let opt_entry = silicon_final_entries.get(&vir_name);
+                    backtranslator.backtranslate_entry(
+                        typ, 
+                        opt_entry, 
+                        vir_name,
+                        silicon_final_entries,
+                    )
+                }
+            }
```

Comment:
This part is almost identical between the pure and impure cases, it can happen after.

Reply:
I refactored this function quite a bit, I realized it was a huge mess and your comment to split it up into multiple definitely makes sense. This was also resolved in the process.